README.md
# Deploy CloudFormation stacks in Gulp

Does what it says on the can.

## Why this package?

The headline benefit over other packages is the surfacing of stack errors:
![Example console error output](https://s3.amazonaws.com/mindhive-package-media/gulp-cf-deploy/error-output-example.png)

Other benefits:

- Lists resources created / updated / deleted
- Errors are displayed ASAP, so you can get to fixing them while the rollback completes
- On failure a URL is provided direct to the stack details in the AWS console for full details
- Parameters can be passed as a hash / plain object (rather than in CloudFormation's verbose format)
- Waits for the stack to be deployed before finishing, so tasks in series can rely on the resources being available
- But does not wait for cleanup of old resources, which should not impact following tasks
- Keeps you updated while waiting for the deployment to complete
- Pipes the outputs of the stack as simplified JSON through the Gulp stream so you can save the outputs or process
  them further
- Defaults to deleting a failed stack creation rather than the CloudFormation default of 'rollback' which then has
  to be manually deleted to try again (however, even when the delete has completed the full details of the
  failed stack are still available at the console URL)

## Requirements

This packages has a peer dependency on Gulp v4, which you can currently
install as `gulp@next`.

## Install

`yarn add --dev gulp-cf-deploy`

Or if you're still in npm world: `npm install --save-dev gulp-cf-deploy`

## API

```js
import cfDeploy from 'gulp-cf-deploy'

gulp.task('deploy:aws', () =>
  gulp
    .src('resources.yaml')
    .pipe(cfDeploy(awsServiceOptions, stackNameOrOptions, parameters))
    .pipe(gulp.dest('build')),
)
```

Will deploy (create or update) the CloudFormation stack defined in `resources.yaml`
and save it's outputs as `build/resources.json`

`awsServiceOptions`: passed to [`new AWS.CloudFormation()`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#constructor-property).
Provide your AWS credentials (if not already set in `AWS.config`) and `region`.

`stackNameOrOptions`: passed to [`createStack()`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#createStack-property)
or [`updateStack()`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#updateStack-property).
Often all that's needed is the `StackName` in which case you can just pass the name as a string.
In addition:

- `StackName`: defaulted to the source file's 'stem', in the example above the stack will be named `resources`
- `Parameters`: will be supplemented with the 3rd argument `parameters`
- `TemplateBody`: is pulled from the source file's content
- `OnFailure`: set to `DELETE` when creating a stack, otherwise the stack needs to be manually deleted to try again
  (even when deleted the stack can be inspected in the AWS Console for 30 days)
- `Capabilities`: Needs to be set if creating IAM resources, see the
  [SDK docs](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CloudFormation.html#createStack-property).

`parameters`: a hash / plain object of parameters that is merged into `Parameters` (if any) of `stackNameOrOptions`.
For example: `{ Foo: 123, ... }` becomes `[{ ParameterKey: 'Foo', ParameterValue: 123 }, ...]`

Output Vinyl file: The `Outputs` of the stack is simplified into a hash / plain object.
For example `[{ OutputKey: 'Foo', OutputValue: 123 }, ...]` becomes `{ Foo: 123, ... }`.
The stream output is a file with the same properties as the source file expect:

- `contents`: JSON of the simplified outputs, pipe through `gulp-json-editor` to easily modify
- `data`: the simplified outputs object
- `extname`: extension is set to `json`
- Name is unchanged: pipe through `gulp-rename` if you want to change it

## Post-processing example

If you are creating IAM access keys then there are a couple of considerations.

1. CloudFormation will only output the SecretAccessKey when the key is first created (or regenerated)
2. When doing anything with IAM users you need to specify so in the stack options

```yaml

---
Resources:
  fooBucket:
    Type: 'AWS::S3::Bucket'
    Properties:
      BucketName: 'bucket-of-foo'
  bobUser:
    Type: 'AWS::IAM::User'
    Properties:
      Policies: ...
  bobAccessKey:
    Type: 'AWS::IAM::AccessKey'
    Properties:
      Serial: 1
      UserName: !Ref bobUser
Outputs:
  fooBucketArn:
    Value: !Sub 'arn:aws:s3:::${fooBucket}'
  bobAccessKey:
    Value: !Ref bobAccessKey
  bobSecretKey:
    Value: !GetAtt bobAccessKey.SecretAccessKey
```

CloudFormation will output `fooBucketArn` and `bobAccessKey` each deployment.
But `bobSecretKey` will only be output when the keys are first created, or regenerated
(for example, if the `Serial` was changed). Also, for security, you would want to be managing
you secret and access keys separately from other resource information (such as `fooBucketArn`).

You can handle this as follows:

```js
import gulp from 'gulp'
import clone from 'gulp-clone'
import filter from 'gulp-filter'
import rename from 'gulp-rename'
import streamToPromise from 'stream-to-promise'
import jsonEditor from 'gulp-json-editor'
import omitBy from 'lodash/omitBy'

const region = 'ap-southeast-2'
gulp.task('deploy:aws', () => {
  const cfOutput = gulp.src('deploy/resources.yaml').pipe(
    cfDeploy(
      {
        region,
      },
      {
        Capabilities: ['CAPABILITY_IAM'], // Required because we are creating a user
        StackName: `project-resources`,
      },
    ),
  )
  const secrets = cfOutput
    .pipe(clone())
    .pipe(filter(file => file.data.bobSecretKey)) // Only perform if secret key was output
    .pipe(
      jsonEditor(resources => ({
        accessKeyId: resources.bobAccessKey,
        secretAccessKey: resources.bobSecretKey,
      })),
    )
    .pipe(rename('bob-credentials.json'))
    .pipe(gulp.dest('secrets', { mode: 0o600 }))
  const resources = cfOutput
    .pipe(
      jsonEditor(resources => ({
        ...omitBy(resources, (v, k) => k.endsWith('Key')),
        region,
      })),
    )
    .pipe(gulp.dest('build')) // No need to rename, will be resources.json
  return Promise.all([secrets, resources].map(streamToPromise))
})
```


CHANGELOG.md
# Changelog


## 3.0.0

- Updated peer dependency to aws-sdk version 3.

## 2.2.0

- Added `taskOptions` as 4th parameter. Allowing setting of `logPrefix` to differentiate
  outputs from multiple tasks. Defaults to stack name.

## 2.1.0

- Converted to Typescript.

## 2.0.0

- Changed to relying on Gulp v4 as a peer dependency.


src/deploy.ts
import {
  CloudFormationClient,
  CloudFormationClientConfig,
  CreateStackCommand,
  CreateStackInput,
  DescribeStackEventsCommand, 
  DescribeStacksCommand,
  Parameter,
  Stack,
  StackEvent,
  UpdateStackCommand,
  UpdateStackInput,
} from '@aws-sdk/client-cloudformation';
import through2 from 'through2'
import PluginError from 'plugin-error'
import log from 'fancy-log'
import colors from 'ansi-colors'
import fromPairs from 'lodash/fromPairs'
import takeWhile from 'lodash/takeWhile'
import startsWith from 'lodash/startsWith'
import endsWith from 'lodash/endsWith'
import includes from 'lodash/includes'

const MAX_REPORT_SECONDS = 15

const error = (err: string | Error) =>
  new PluginError('cf-deploy', err, { showProperties: typeof err !== 'string' })

const isInProgress = (state: Stack) =>
  endsWith(state.StackStatus, '_IN_PROGRESS')

const isCleanupInProgress = (state: Stack) =>
  endsWith(state.StackStatus, '_CLEANUP_IN_PROGRESS')

const isInitialStackEvent = (event: StackEvent, StackId: string) =>
  event.PhysicalResourceId === StackId &&
  ['CREATE_IN_PROGRESS', 'UPDATE_IN_PROGRESS'].some(
    s => s === event.ResourceStatus,
  )

const isFailedStackEvent = (event: StackEvent) =>
  endsWith(event.ResourceStatus, 'FAILED')

const stackStateIndicatesFailure = (state: Stack) =>
  ['ROLLBACK', 'DELETE', 'FAILED'].some(s => includes(state.StackStatus, s))

const isDeploySuccessful = (state: Stack) =>
  includes(state.StackStatus, 'COMPLETE') &&
  !includes(state.StackStatus, 'ROLLBACK') &&
  ['CREATE', 'UPDATE'].some(s => startsWith(state.StackStatus, s))

const simplifiedOutput = (state: Stack) =>
  fromPairs(
    state.Outputs!.map(({ OutputKey, OutputValue }) => [
      OutputKey,
      OutputValue,
    ]),
  )

interface IServiceOptions extends CloudFormationClientConfig {
  region: string // We require region
}

interface ITaskOptions {
  logPrefix?: string
}

type StackInput = CreateStackInput | UpdateStackInput

export type ParameterValue = string | number

export type HashParameters = Record<string, ParameterValue>

const parameterValue = (param: ParameterValue): string => String(param)

// noinspection JSUnusedGlobalSymbols
export default (
  serviceOptions: IServiceOptions,
  stackNameOrOptions: string | StackInput | undefined,
  parameters: HashParameters = {},
  options: ITaskOptions = {},
) =>
  through2.obj(async (file, enc, done) => {
    const deploy = async () => {
      const stackOptions: StackInput =
        typeof stackNameOrOptions === 'string'
          ? { StackName: stackNameOrOptions }
          : typeof stackNameOrOptions === 'undefined'
          ? { StackName: file.stem }
          : stackNameOrOptions

      if (file.isNull()) {
        return file
      }
      if (!file.isBuffer()) {
        throw error(
          'Can only handle buffered files, pipe through vinyl-buffer beforehand',
        )
      }
      const stackName = stackOptions.StackName
      const logPrefix = options.logPrefix || stackName
      const cfn = new CloudFormationClient({
        apiVersion: '2010-05-15',
        ...serviceOptions,
      })

      const retrieveStackEvents = async (stackId: string) => {
        const command = new DescribeStackEventsCommand({ StackName: stackId });
        const eventsResult = await cfn
          .send(command)
        return takeWhile(
          eventsResult.StackEvents,
          e => !isInitialStackEvent(e, stackId),
        ).reverse() // Change to chronological order
      }

      const logFailures = (stackEvents: StackEvent[]) => {
        stackEvents
          .filter(e => isFailedStackEvent(e))
          .forEach(e =>
            log(
              `${logPrefix}.${e.LogicalResourceId}:`,
              colors.dim(`(${e.ResourceStatus})`),
              colors.red(
                e.ResourceStatusReason || e.ResourceStatus || '<unknown>',
              ),
            ),
          )
      }

      const buildParameters = (): Parameter [] => [
        ...(stackOptions.Parameters || []),
        ...Object.entries(parameters).map(([k, v]) => ({
          ParameterKey: k,
          ParameterValue: parameterValue(v),
        })),
      ]

      const completedState = async ({
        StackId,
        reportEvents,
        waitForCleanup,
      }: {
        StackId: string
        reportEvents: boolean
        waitForCleanup: boolean
      }) => {
        let haveReportedFailures = false
        const reportedLogicalIds: string[] = []
        let state
        let delaySeconds = 2
        const command = new DescribeStacksCommand({ StackName: StackId });
        for (;;) {
          const describeResult = await cfn
            .send(command)
          state = describeResult.Stacks![0]
          const completed =
            !isInProgress(state) ||
            (!waitForCleanup && isCleanupInProgress(state))
          if (reportEvents) {
            const stackEvents = await retrieveStackEvents(StackId)
            stackEvents
              .filter(
                e => e.LogicalResourceId && e.PhysicalResourceId !== StackId,
              )
              .forEach(e => {
                if (!includes(reportedLogicalIds, e.LogicalResourceId)) {
                  const prefix = `${logPrefix}.${e.LogicalResourceId}`
                  if (startsWith(e.ResourceStatus, 'CREATE')) {
                    log(`${prefix}: ${colors.green('✔ creating')}`)
                  } else if (startsWith(e.ResourceStatus, 'UPDATE')) {
                    log(`${prefix}: ✱ updating`)
                  } else if (startsWith(e.ResourceStatus, 'DELETE')) {
                    log(`${prefix}: ${colors.dim('✘ deleting')}`)
                  }
                  reportedLogicalIds.push(e.LogicalResourceId!)
                }
              })
            if (!haveReportedFailures && stackStateIndicatesFailure(state)) {
              logFailures(stackEvents)
              log(
                'See full details at:',
                colors.blue.underline(
                  `https://${
                    serviceOptions.region
                  }.console.aws.amazon.com/cloudformation/home?region=${
                    serviceOptions.region
                  }#/stack/detail?stackId=${encodeURIComponent(StackId)}`,
                ),
              )
              haveReportedFailures = true
            }
          }
          log(
            `${logPrefix}: ${state.StackStatus}`,
            completed
              ? ''
              : colors.dim(`checking again in ${delaySeconds}s...`),
          )
          if (completed) {
            return state
          }
          await new Promise(resolve => setTimeout(resolve, delaySeconds * 1000)) // eslint-disable-line
          delaySeconds = Math.min(delaySeconds * 2, MAX_REPORT_SECONDS)
        }
      }

      let initialState: Stack | undefined
      try {
        const command = new DescribeStacksCommand({ StackName: stackName });
        const describeResult = await cfn
          .send(command)
        initialState = describeResult.Stacks![0]
      } catch (e) {
        if (e.code !== 'ValidationError') {
          throw e
        }
      }
      if (initialState) {
        if (isInProgress(initialState)) {
          log(`${logPrefix}: waiting for previous operation to complete`)
          initialState = await completedState({
            StackId: initialState.StackId!,
            reportEvents: false,
            waitForCleanup: true,
          })
          log(`${logPrefix}: starting deployment`)
        }
      } else {
        log(`${logPrefix}: creating new stack`)
      }
      const updating = initialState != null
      const deployParams: StackInput = {
        ...(updating ? {} : { OnFailure: 'DELETE' }), // Put this first so can be overridden by passed params
        ...stackOptions,
        Parameters: buildParameters(),
        TemplateBody: file.contents.toString(enc),
      }
      let resultState
      let skipped = false
      try {
        const command = updating ? new UpdateStackCommand(deployParams) : new CreateStackCommand(deployParams)
        const deployResult = await cfn.send(command)
        resultState = await completedState({
          StackId: deployResult.StackId!,
          reportEvents: true,
          waitForCleanup: false,
        })
      } catch (e) {
        if (
          updating &&
          e.code === 'ValidationError' &&
          e.message === 'No updates are to be performed.'
        ) {
          // CloudFormation will only update if *resources* will change, for example: changes to Outputs don't count
          resultState = initialState! // Must exist since we're updating
          skipped = true
        } else {
          throw e
        }
      }
      if (!skipped && !isDeploySuccessful(resultState)) {
        if (endsWith(resultState.StackStatus, '_FAILED')) {
          logFailures(await retrieveStackEvents(resultState.StackId!))
        }
        throw error(
          `Deploying ${stackName} failed (${resultState.StackStatusReason ||
            resultState.StackStatus})`,
        )
      }
      const output = simplifiedOutput(resultState)
      const outputFile = file.clone({ contents: false })
      outputFile.contents = Buffer.from(JSON.stringify(output, null, 2), 'utf8')
      outputFile.data = output
      outputFile.extname = 'json'
      return outputFile
    }

    try {
      done(null, await deploy())
    } catch (e) {
      if (e instanceof PluginError) {
        done(e)
      } else {
        done(error(e))
      }
    }
  })


src/deploy.js
import CloudFormation from 'aws-sdk/clients/cloudformation'
import through2 from 'through2'
import PluginError from 'plugin-error'
import log from 'fancy-log'
import colors from 'ansi-colors'
import fromPairs from 'lodash/fromPairs'
import toPairs from 'lodash/toPairs'
import takeWhile from 'lodash/takeWhile'
import startsWith from 'lodash/startsWith'
import endsWith from 'lodash/endsWith'
import includes from 'lodash/includes'

const MAX_REPORT_SECONDS = 15

class Err extends PluginError {
  constructor(message) {
    super('cf-deploy', message, {
      showProperties: typeof message === 'object',
    })
  }
}

const isInProgress = state => endsWith(state.StackStatus, '_IN_PROGRESS')

const isCleanupInProgress = state =>
  endsWith(state.StackStatus, '_CLEANUP_IN_PROGRESS')

const isInitialStackEvent = (event, StackId) =>
  event.PhysicalResourceId === StackId &&
  ['CREATE_IN_PROGRESS', 'UPDATE_IN_PROGRESS'].some(
    s => s === event.ResourceStatus,
  )

const isFailedStackEvent = event => endsWith(event.ResourceStatus, 'FAILED')

const stackStateIndicatesFailure = state =>
  ['ROLLBACK', 'DELETE', 'FAILED'].some(s => includes(state.StackStatus, s))

const isDeploySuccessful = state =>
  includes(state.StackStatus, 'COMPLETE') &&
  !includes(state.StackStatus, 'ROLLBACK') &&
  ['CREATE', 'UPDATE'].some(s => startsWith(state.StackStatus, s))

const simplifiedOutput = state =>
  fromPairs(
    state.Outputs.map(({ OutputKey, OutputValue }) => [OutputKey, OutputValue]),
  )

export default (serviceOptions = {}, stackNameOrOptions, parameters = {}) =>
  through2.obj(async (file, enc, done) => {
    const stackOptions =
      typeof stackNameOrOptions === 'string'
        ? { StackName: stackNameOrOptions }
        : stackNameOrOptions || {}

    const deploy = async () => {
      if (file.isNull()) {
        return file
      }
      if (!file.isBuffer()) {
        throw new Err(
          'Can only handle buffered files, pipe through vinyl-buffer beforehand',
        )
      }
      const StackName = stackOptions.StackName || file.stem
      const cfn = new CloudFormation({
        apiVersion: '2010-05-15',
        ...serviceOptions,
      })

      const retrieveStackEvents = async ({ StackId }) => {
        const eventsResult = await cfn
          .describeStackEvents({ StackName: StackId })
          .promise()
        return takeWhile(
          eventsResult.StackEvents,
          e => !isInitialStackEvent(e, StackId),
        ).reverse() // Change to chronological order
      }

      const logFailures = stackEvents => {
        stackEvents
          .filter(e => isFailedStackEvent(e))
          .forEach(e =>
            log(
              `${StackName}.${e.LogicalResourceId}:`,
              colors.dim(`(${e.ResourceStatus})`),
              colors.red(e.ResourceStatusReason || e.ResourceStatus),
            ),
          )
      }

      const completedState = async ({
        StackId,
        reportEvents,
        waitForCleanup,
      }) => {
        let haveReportedFailures = false
        const reportedLogicalIds = []
        let state
        let delaySeconds = 2
        for (;;) {
          const describeResult = await cfn
            .describeStacks({ StackName: StackId })
            .promise()
          state = describeResult.Stacks[0]
          const completed =
            !isInProgress(state) ||
            (!waitForCleanup && isCleanupInProgress(state))
          if (reportEvents) {
            const stackEvents = await retrieveStackEvents({ StackId })
            stackEvents
              .filter(
                e => e.LogicalResourceId && e.PhysicalResourceId !== StackId,
              )
              .forEach(e => {
                if (!includes(reportedLogicalIds, e.LogicalResourceId)) {
                  const prefix = `${StackName}.${e.LogicalResourceId}`
                  if (startsWith(e.ResourceStatus, 'CREATE')) {
                    log(`${prefix}: ${colors.green('✔ creating')}`)
                  } else if (startsWith(e.ResourceStatus, 'UPDATE')) {
                    log(`${prefix}: ✱ updating`)
                  } else if (startsWith(e.ResourceStatus, 'DELETE')) {
                    log(`${prefix}: ${colors.dim('✘ deleting')}`)
                  }
                  reportedLogicalIds.push(e.LogicalResourceId)
                }
              })
            if (!haveReportedFailures && stackStateIndicatesFailure(state)) {
              logFailures(stackEvents)
              log(
                'See full details at:',
                colors.blue.underline(
                  `https://${
                    serviceOptions.region
                  }.console.aws.amazon.com/cloudformation/home?region=${
                    serviceOptions.region
                  }#/stack/detail?stackId=${encodeURIComponent(StackId)}`,
                ),
              )
              haveReportedFailures = true
            }
          }
          log(
            `${StackName}: ${state.StackStatus}`,
            completed
              ? ''
              : colors.dim(`checking again in ${delaySeconds}s...`),
          )
          if (completed) {
            return state
          }
          await new Promise(resolve => setTimeout(resolve, delaySeconds * 1000)) // eslint-disable-line
          delaySeconds = Math.min(delaySeconds * 2, MAX_REPORT_SECONDS)
        }
      }

      let initialState
      try {
        const describeResult = await cfn.describeStacks({ StackName }).promise()
        initialState = describeResult.Stacks[0]
      } catch (e) {
        if (e.code !== 'ValidationError') {
          throw e
        }
      }
      if (initialState) {
        if (isInProgress(initialState)) {
          log(`${StackName}: waiting for previous operation to complete`)
          initialState = await completedState({
            StackId: initialState.StackId,
            reportEvents: false,
            waitForCleanup: true,
          })
          log(`${StackName}: starting deployment`)
        }
      } else {
        log(`${StackName}: creating new stack`)
      }
      const updating = initialState != null
      const deployParams = {
        ...(updating ? {} : { OnFailure: 'DELETE' }), // Put this first so can be overridden by passed params
        ...stackOptions,
        Parameters: [
          ...(stackOptions.Parameters || []),
          ...toPairs(parameters).map(([k, v]) => ({
            ParameterKey: k,
            ParameterValue: v,
          })),
        ],
        StackName,
        TemplateBody: file.contents.toString(enc),
      }
      let resultState
      let skipped = false
      try {
        const deployResult = await cfn[
          updating ? 'updateStack' : 'createStack'
        ](deployParams).promise()
        resultState = await completedState({
          StackId: deployResult.StackId,
          reportEvents: true,
          waitForCleanup: false,
        })
      } catch (e) {
        if (
          updating &&
          e.code === 'ValidationError' &&
          e.message === 'No updates are to be performed.'
        ) {
          // CloudFormation will only update if *resources* will change, for example: changes to Outputs don't count
          resultState = initialState
          skipped = true
        } else {
          throw e
        }
      }
      if (!skipped && !isDeploySuccessful(resultState)) {
        if (endsWith(resultState.StackStatus, '_FAILED')) {
          logFailures(
            await retrieveStackEvents({ StackId: resultState.StackId }),
          )
        }
        throw new Err(
          `Deploying ${StackName} failed (${resultState.StackStatusReason ||
            resultState.StackStatus})`,
        )
      }
      const output = simplifiedOutput(resultState)
      const outputFile = file.clone({ contents: false })
      outputFile.contents = Buffer.from(JSON.stringify(output, null, 2), 'utf8')
      outputFile.data = output
      outputFile.extname = 'json'
      return outputFile
    }

    try {
      done(null, await deploy())
    } catch (e) {
      if (e instanceof PluginError) {
        done(e)
      } else {
        done(new Err(e))
      }
    }
  })
